### PR TITLE
setup jupyterhub config

### PR DESCRIPTION
**Pull request by Tianhao-Gu**

### CODE REVIEW

File: scripts/spawn_notebook.sh [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+
+USERNAME=${JUPYTERHUB_USER}
+
+echo "Starting Jupyter Notebook for user: $USERNAME"
+cd $JUPYTERHUB_USER_HOME/$USERNAME
```
Comment (Tianhao-Gu) [inline scripts/spawn_notebook.sh:6]: User home dir will also be created in `custom_spawner.py` in future PRs.

File: Dockerfile [hunk 2]
```diff
@@ -82,6 +98,10 @@ ENV CDM_SHARED_DIR=/cdm_shared_workspace
 RUN mkdir -p ${CDM_SHARED_DIR} && chmod -R 777 ${CDM_SHARED_DIR}
 RUN chown -R spark_user:spark $CDM_SHARED_DIR
 
+# Allow spark_user to use sudo without a password
```
Comment (Tianhao-Gu) [inline Dockerfile:101]: Configuring JupyterHub with a non-root user is a lot challenging, as this user must manage other system users (home dir, virtual env, etc). JupyterHub requires the presence of system users for its operations.

File: config/jupyterhub_config.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+import os
+
+from jupyterhub_config.custom_spawner import VirtualEnvSpawner
+
+c = get_config()
+
+# Set the authenticator class
+# TODO: Change the authenticator class to a secure one (e.g. GitHubOAuthenticator)
+c.JupyterHub.authenticator_class = 'jupyterhub.auth.DummyAuthenticator'
+c.Authenticator.allowed_users = {'spark_user', 'test_user1', 'test_user2'}
```
Comment (Tianhao-Gu) [inline config/jupyterhub_config.py:15]: Non-existent system users (`test_user1`, `test_user2`) will be automatically created later in `custom_spawner.py`.

Comment (Tianhao-Gu) [reply config/jupyterhub_config.py:15]: Eventually I want to switch to OAuth and phase out user/password auth completely. Maybe only allow admin users to use password.

Comment (MrCreosote) [reply config/jupyterhub_config.py:15]: Presumably long term we'd swap out all the auth systems for KBase auth

Comment (Tianhao-Gu) [reply config/jupyterhub_config.py:15]: yea. That's the plan to use KBase auth eventually.

File: Dockerfile [hunk 4]
```diff
@@ -82,6 +98,10 @@ ENV CDM_SHARED_DIR=/cdm_shared_workspace
...
+# TODO: use `sudospawner` in JupyterHub to avoid this (https://jupyterhub.readthedocs.io/en/stable/howto/configuration/config-sudo.html)
+RUN echo "spark_user ALL=(ALL) NOPASSWD: ALL" >> /etc/sudoers
```
Comment (MrCreosote) [inline Dockerfile:103]: How do you ensure users can't access this account?

File: config/jupyterhub_config.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
+"""
+This is the JupyterHub configuration file. It is used to configure the JupyterHub server.
+Refer to the JupyterHub documentation for more information:
+https://jupyterhub.readthedocs.io/en/latest/tutorial/getting-started/config-basics.html
+"""
+import os
+
+from jupyterhub_config.custom_spawner import VirtualEnvSpawner
+
+c = get_config()
```
Comment (MrCreosote) [inline config/jupyterhub_config.py:10]: Where does the `get_config` name come from? I'd expect this to throw a NameError

File: config/jupyterhub_config.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+# Set the authenticator class
+# TODO: Change the authenticator class to a secure one (e.g. GitHubOAuthenticator)
+c.JupyterHub.authenticator_class = 'jupyterhub.auth.DummyAuthenticator'
+c.Authenticator.allowed_users = {'spark_user', 'test_user1', 'test_user2'}
+c.DummyAuthenticator.password = os.environ['JUPYTERHUB_ADMIN_PASSWORD']
```
Comment (MrCreosote) [inline config/jupyterhub_config.py:16]: is there an authenticator that allows setting a password per user?

File: scripts/spawn_notebook.sh [hunk 7]
```diff
@@ -0,0 +1,9 @@
...
+
+# Start the notebook server with current user
+exec jupyterhub-singleuser "$@"
```
Comment (MrCreosote) [inline scripts/spawn_notebook.sh:9]: This looks like a shell injection vulnerability

**Review by MrCreosote — COMMENTED** [review]
I don't really understand enough about how jupyterhub works to properly review this. What prevents user A from reading from or writing to user B's environment?

File: scripts/spawn_notebook.sh (hunk 7 above)
Comment (Tianhao-Gu) [reply scripts/spawn_notebook.sh]: 👍

File: config/jupyterhub_config.py (hunk 6 above)
Comment (Tianhao-Gu) [reply config/jupyterhub_config.py:16]: I think there is but I want to explore that in the future.

File: config/jupyterhub_config.py (hunk 5 above)
Comment (Tianhao-Gu) [reply config/jupyterhub_config.py:10]: It's imported automatically by jupyterhub.

File: config/jupyterhub_config.py (hunk 6 above)
Comment (MrCreosote) [reply config/jupyterhub_config.py:16]: :+1:

File: config/jupyterhub_config.py (hunk 5 above)
Comment (MrCreosote) [reply config/jupyterhub_config.py:10]: E.g. it monkey patches the config file?

File: scripts/spawn_notebook.sh (hunk 7 above)
Comment (MrCreosote) [reply scripts/spawn_notebook.sh]: :+1:

File: Dockerfile (hunk 4 above)
Comment (Tianhao-Gu) [reply Dockerfile:103]: users logged in as individual system user. I don't know they can have access to spark_user.

File: Dockerfile [hunk 8]
```diff
@@ -42,22 +54,26 @@ RUN pipenv sync --system
 
 RUN chown -R spark_user:spark /opt/bitnami
 
-# Set up Jupyter Lab directories
+# Set up JupyterLab directories
```
[the remark this answers is not in the document]
Comment (Tianhao-Gu) [reply Dockerfile:57]: yea. I think so.

[the remark this answers is not in the document]
Comment (MrCreosote) [reply Dockerfile:57]: :+1:

File: Dockerfile (hunk 4 above)
Comment (MrCreosote) [reply Dockerfile:103]: So you have to make a user account for each user?

**Comment by Tianhao-Gu** [comment]
> I don't really understand enough about how jupyterhub works to properly review this. What prevents user A from reading from or writing to user B's environment?

Upon logging in, we need to ensure user's home directory is properly created and permissioned which will be done by `VirtualEnvSpawner` class.

File: Dockerfile (hunk 4 above)
Comment (Tianhao-Gu) [reply Dockerfile:103]: Yes. Jupyterhub operates only on system users.

Comment (MrCreosote) [reply Dockerfile:103]: Ok, so what the spawner is going to do it make a new user if the user & set up home if it doesn't exist, and rely on the linux permissions to make sure they can't read other user's envs?

File: config/jupyterhub_config.py (hunk 5 above)
Comment (Tianhao-Gu) [reply config/jupyterhub_config.py:10]: hmmm not exactly sure. But it's ready to use by Jupyterhub automatically. 
https://jupyterhub.readthedocs.io/en/stable/reference/config-reference.html

File: Dockerfile (hunk 4 above)
Comment (Tianhao-Gu) [reply Dockerfile:103]: Yea. Will implement those in the future PRs.

File: config/jupyterhub_config.py (hunk 5 above)
Comment (MrCreosote) [reply config/jupyterhub_config.py:10]: can you link to that in the config as well as the link you already have

File: Dockerfile (hunk 4 above)
Comment (MrCreosote) [reply Dockerfile:103]: :+1:

File: config/jupyterhub_config.py (hunk 5 above)
Comment (Tianhao-Gu) [reply config/jupyterhub_config.py:11]: 👍

Comment (MrCreosote) [reply config/jupyterhub_config.py:11]: :+1:

**Review by MrCreosote — APPROVED** [review]

**Pull request merged**